src/parser.ts

```diff
@@ -413,45 +412,31 @@ class RegExpParserState {
         strings: boolean,
     ): void {
         const parent = this._node
-        if (
-            (parent.type !== "Alternative" &&
-                parent.type !== "CharacterClass") ||
-            (strings && (negate || value))
-        ) {
+        if (parent.type !== "Alternative" && parent.type !== "CharacterClass") {
             throw new Error("UnknownError")
         }
 
         const base = {
             type: "CharacterSet",
-            parent,
             start,
             end,
             raw: this.source.slice(start, end),
             kind,
-            strings,
             key,
         } as const
-        const node: UnicodePropertyCharacterSet = strings
-            ? {
-                  ...base,
-                  value: null,
-                  negate: false,
-                  strings: true,
-              }
-            : {
-                  ...base,
-                  value,
-                  negate,
-                  strings: false,
-              }
 
-        if (node.strings) {
-            if (parent.type === "CharacterClass" && !parent.unicodeSets) {
+        if (strings) {
+            if (
+                (parent.type === "CharacterClass" && !parent.unicodeSets) ||
+                negate ||
+                value !== null
+            ) {
                 throw new Error("UnknownError")
             }
-            parent.elements.push(node)
+
+            parent.elements.push({ ...base, parent, strings, value, negate })
         } else {
-            parent.elements.push(node)
+            parent.elements.push({ ...base, parent, strings, value, negate })
         }
```

Comment:
What about refactoring this to

```suggestion
        if (strings && ((parent.type === "CharacterClass" && !parent.unicodeSets) ||
                negate ||
                value !== null)) {
            throw new Error("UnknownError")
        } else {
            parent.elements.push({ ...base, parent, strings, value, negate })
        }
```

Reply:
Doesn't type check. TS isn't smart enough for dependent types. Branching on `strings` if unfortunately necessary.

![image](https://github.com/eslint-community/regexpp/assets/20878432/18a44549-bc71-48cf-89c1-5f8b996a35eb)

Reply:
The `else` case isn't even necessary as the if-case will throw

Not sure if that makes TS happy though

Reply:
It certainly won't. TS's control flow analysis is smart enough to figure out that `if (cond) { throw X } else { foo() }` is equivalent to `if (cond) { throw X } foo()`.